### PR TITLE
Add checkpoint parameter to on_save_checkpoint

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

See title

```python
class MyCallback(pl.Callback):
    def on_save_checkpoint(trainer, pl_module, checkpoint: ADDED):
       # the checkpoint parameter is the current checkpoint state. this function
       # should still return the callback state.
       ...
```

Necessary because the `ModelPruning` callback needs to modify the saved checkpoint model state. See #6073 

## Backwards compatibility

Inspect the target callback `on_save_checkpoint` signature and if it only has two arguments print a `DeprecationWarning` and call it without the `checkpoint`

## Before submitting
- [n/a] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

File: pytorch_lightning/callbacks/base.py [hunk 1]
```diff
@@ -177,12 +177,26 @@ def on_keyboard_interrupt(self, trainer, pl_module: LightningModule) -> None:
...
+        Returns:
+            The callback state.
+        """
         pass
 
-    def on_load_checkpoint(self, checkpointed_state) -> None:
-        """Called when loading a model checkpoint, use to reload state."""
+    def on_load_checkpoint(self, callback_state: Dict[str, Any]) -> None:
+        """Called when loading a model checkpoint, use to reload state.
+
```
Comment (carmocca) [inline pytorch_lightning/callbacks/base.py:196]:
```suggestion

```

**Comment by rohitgr7** [comment]
are we allowed to update the existing hooks??
https://github.com/PyTorchLightning/pytorch-lightning/pull/6059#issuecomment-781574969
https://github.com/PyTorchLightning/pytorch-lightning/pull/6059#issuecomment-781592005

**Comment by carmocca** [comment]
> are we allowed to update the existing hooks??

Why not? Especially if it's necessary for progress. We just have to be careful about backwards compatibility.

**Comment by SeanNaren** [comment]
> are we allowed to update the existing hooks??
> [#6059 (comment)](https://github.com/PyTorchLightning/pytorch-lightning/pull/6059#issuecomment-781574969)
> [#6059 (comment)](https://github.com/PyTorchLightning/pytorch-lightning/pull/6059#issuecomment-781592005)

A last minute change to the API for the release, I didn't want to add too much additional logic especially to existing APIs. @carmocca suggested fix by inspection will work fine!

**Comment by rohitgr7** [comment]
> > are we allowed to update the existing hooks??
> > [#6059 (comment)](https://github.com/PyTorchLightning/pytorch-lightning/pull/6059#issuecomment-781574969)
> > [#6059 (comment)](https://github.com/PyTorchLightning/pytorch-lightning/pull/6059#issuecomment-781592005)
> 
> A last minute change to the API for the release, I didn't want to add too much additional logic especially to existing APIs. @carmocca suggested fix by inspection will work fine!

@SeanNaren so can we add it now, and of course, complete it with the correct dataloader_idx?

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/trainer/callback_hook.py [hunk 2]
```diff
@@ -197,14 +199,29 @@ def on_keyboard_interrupt(self):
         for callback in self.callbacks:
             callback.on_keyboard_interrupt(self, self.lightning_module)
 
-    def on_save_checkpoint(self):
+    @staticmethod
+    def __is_old_signature(fn: Callable) -> bool:
```
Comment (rohitgr7) [inline pytorch_lightning/trainer/callback_hook.py:203]: can we make this a normal function that can be used anywhere, just in-case we need to deprecate more signatures? it can take old parameters count (`2`) as input.

**Review by rohitgr7 — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !

File: CHANGELOG.md [hunk 3]
```diff
@@ -10,6 +10,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 ### Added
 
 
+- Add `checkpoint` parameter to callback's `on_save_checkpoint` hook ([#6072](https://github.com/PyTorchLightning/pytorch-lightning/pull/6072))
```
Comment (kaushikb11) [inline CHANGELOG.md:13]:
```suggestion
- Added `checkpoint` parameter to callback's `on_save_checkpoint` hook ([#6072](https://github.com/PyTorchLightning/pytorch-lightning/pull/6072))
```

File: pytorch_lightning/trainer/callback_hook.py (hunk 2 above)
Comment (carmocca) [reply pytorch_lightning/trainer/callback_hook.py:203]: I don't think there is a point right now, as we don't know how useful it will be for other deprecations.

If a future PR could use it, the author can take this and move it outside if necessary.

File: pytorch_lightning/callbacks/base.py [hunk 4]
```diff
@@ -177,12 +177,25 @@ def on_keyboard_interrupt(self, trainer, pl_module: LightningModule) -> None:
...
+        Returns:
+            The callback state.
+        """
         pass
 
-    def on_load_checkpoint(self, checkpointed_state) -> None:
-        """Called when loading a model checkpoint, use to reload state."""
+    def on_load_checkpoint(self, callback_state: Dict[str, Any]) -> None:
+        """Called when loading a model checkpoint, use to reload state.
+        Args:
```
Comment (carmocca) [inline pytorch_lightning/callbacks/base.py:196]:
```suggestion

        Args:
```

File: pytorch_lightning/trainer/callback_hook.py [hunk 5]
```diff
@@ -197,14 +199,29 @@ def on_keyboard_interrupt(self):
...
+            return True
+        return False
+
+    def on_save_checkpoint(self, checkpoint: Dict[str, Any]) -> Dict[Type, dict]:
         """Called when saving a model checkpoint."""
         callback_states = {}
         for callback in self.callbacks:
-            callback_class = type(callback)
-            state = callback.on_save_checkpoint(self, self.lightning_module)
+            if self.__is_old_signature(callback.on_save_checkpoint):
```
Comment (kaushikb11) [inline pytorch_lightning/trainer/callback_hook.py:213]: @carmocca staticmethod and calling by self?

**Review by williamFalcon — APPROVED** [review]
